File: Count Nodes Equal to Average of Subtree - Leetcode 2265.py
class Solution:
    def averageOfSubtree(self, root: TreeNode) -> int:
        num_nodes = [0]

        def dfs(root):
            if not root:
                return (0, 0)
            
            N_left, summ_left = dfs(root.left)
            N_right, summ_right = dfs(root.right)

            N = 1 + N_left + N_right
            summ = root.val + summ_left + summ_right
            avg = summ // N

            if root.val == avg:
                num_nodes[0] += 1

            return (N, summ)
        
        dfs(root)
        return num_nodes[0]
        # Time: O(N)
        # Space: O(N)
        
    

File: Count Nodes Equal to Average of Subtree - Leetcode 2265/Count Nodes Equal to Average of Subtree - Leetcode 2265.cpp
#include <algorithm>
using namespace std;

class Solution {
private:
    int numNodes = 0;

public:
    int averageOfSubtree(TreeNode* root) {
        dfs(root);
        return numNodes;
    }

    // Returns a pair: {number of nodes, sum of values}
    pair<int, int> dfs(TreeNode* root) {
        if (!root) return {0, 0};

        auto left = dfs(root->left);
        auto right = dfs(root->right);

        int nodeCount = 1 + left.first + right.first;
        int sum = root->val + left.second + right.second;
        int avg = sum / nodeCount;

        if (root->val == avg) numNodes++;

        return {nodeCount, sum};
    }
};
